packages/promotion/src/services/promotion-module.ts

```diff
@@ -90,6 +91,120 @@ export default class PromotionModuleService<
     return joinerConfig
   }
 
+  @InjectManager("baseRepository_")
+  async registerUsage(
+    computedActions: PromotionTypes.ComputeActions[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<void> {
+    const promotionCodes = computedActions
+      .map((computedAction) => computedAction.code)
+      .filter(Boolean)
+
+    const promotionCodeCampaignBudgetMap = new Map<
+      string,
+      UpdateCampaignBudgetDTO
+    >()
+    const promotionCodeUsageMap = new Map<string, boolean>()
+
+    const existingPromotions = await this.list(
+      { code: promotionCodes },
+      { relations: ["application_method", "campaign", "campaign.budget"] },
+      sharedContext
+    )
+
+    const existingPromotionsMap = new Map<string, PromotionTypes.PromotionDTO>(
+      existingPromotions.map((promotion) => [promotion.code!, promotion])
+    )
+
+    for (let computedAction of computedActions) {
+      if (!ComputeActionUtils.canRegisterUsage(computedAction)) {
+        continue
+      }
+
+      computedAction = computedAction as PromotionTypes.UsageComputedActions
+      const promotion = existingPromotionsMap.get(computedAction.code)
+
+      if (!promotion) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          `Promotion with code ${computedAction.code} not found`
+        )
+      }
+
+      const campaignBudget = promotion.campaign?.budget
+
+      if (!campaignBudget) {
+        continue
+      }
+
+      if (campaignBudget.type === CampaignBudgetType.SPEND) {
+        const campaignBudgetData = promotionCodeCampaignBudgetMap.get(
+          campaignBudget.id
+        ) || { id: campaignBudget.id, used: campaignBudget.used || 0 }
+
+        campaignBudgetData.used =
+          (campaignBudgetData.used || 0) + computedAction.amount
+
+        if (
+          campaignBudget.limit &&
+          campaignBudgetData.used > campaignBudget.limit
+        ) {
+          throw new MedusaError(
```

Comment:
thought(non-blocking): We need to think about how this will work in practice. If the usage is registered as a side effect of completing a cart, then this check will not be of much good use. I guess we will need to have a validation check on the current usage upon applying a promotion to a cart.

Reply:
Thats in my next PR here - https://github.com/medusajs/medusa/pull/6095/files

I was a bit hesitant about throwing here as well. But the condition of when a promo is applied and when the cart is completed can also sometimes be days. But i guess a recomputation can be done by the consumer if the cart is older than X amount of time.

Will remove this validation